src/Microsoft.AspNetCore.Mvc.ViewFeatures/ViewFeatures/DefaultHtmlGenerator.cs

```diff
@@ -1472,53 +1452,76 @@ public IHtmlContent GenerateGroupsAndOptions(string optionLabel, IEnumerable<Sel
+                }, currentValues: null));
+            }
+
+            var itemsList = selectList as IList<SelectListItem>;
+            if (itemsList == null)
+            {
+                itemsList = selectList.ToList();
             }
 
             // Group items in the SelectList if requested.
-            // Treat each item with Group == null as a member of a unique group
-            // so they are added according to the original order.
-            var groupedSelectList = selectList.GroupBy<SelectListItem, int>(
-                item => (item.Group == null) ? item.GetHashCode() : item.Group.GetHashCode());
-            foreach (var group in groupedSelectList)
-            {
-                var optGroup = group.First().Group;
-                if (optGroup != null)
+            // The worst case complexity of algorithm is O(number of groups*n). 
+            // If there aren't any groups, it is O(n) where n is number of items in the list.
+            var optionGenerated = new bool[itemsList.Count];
+            for (var i = 0; i < itemsList.Count; i++)
+            {
+                if (!optionGenerated[i])
                 {
-                    var groupBuilder = new TagBuilder("optgroup");
-                    if (optGroup.Name != null)
+                    var item = itemsList[i];
+                    var optGroup = item.Group;
+                    if (optGroup != null)
                     {
-                        groupBuilder.MergeAttribute("label", optGroup.Name);
-                    }
+                        var groupBuilder = new TagBuilder("optgroup");
+                        if (optGroup.Name != null)
+                        {
+                            groupBuilder.MergeAttribute("label", optGroup.Name);
+                        }
 
-                    if (optGroup.Disabled)
-                    {
-                        groupBuilder.MergeAttribute("disabled", "disabled");
-                    }
+                        if (optGroup.Disabled)
+                        {
+                            groupBuilder.MergeAttribute("disabled", "disabled");
+                        }
 
-                    groupBuilder.InnerHtml.AppendLine();
-                    foreach (var item in group)
-                    {
-                        groupBuilder.InnerHtml.AppendLine(GenerateOption(item));
-                    }
+                        groupBuilder.InnerHtml.AppendLine();
 
-                    listItemBuilder.AppendLine(groupBuilder);
-                }
-                else
-                {
-                    foreach (var item in group)
+                        for (var j = i; j < itemsList.Count; j++)
+                        {
+                            var groupItem = itemsList[j];
+
+                            if (!optionGenerated[j] &&
+                                groupItem.Group != null &&
+                                groupItem.Group.GetHashCode() == item.Group.GetHashCode())
```

Comment:
Use `object.ReferenceEquals(optGroup, groupItem.Group)` instead of `groupItem.Group != null && groupItem.Group.GetHashCode() == item.Group.GetHashCode()`. For one thing, you've got `optGroup` and don't need to reference `item.Group` again. For another, the previous use of `GetHashCode()` was a hack to maintain placement of items not in groups; no need for that now. See also the comments at the top of [`SelectListGroup`](https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNetCore.Mvc.ViewFeatures/Rendering/SelectListGroup.cs).